localstack-core/localstack/services/apigateway/next_gen/execute_api/context.py

```diff
@@ -1,22 +1,67 @@
-from typing import Optional
+from http import HTTPMethod
+from typing import Optional, TypedDict
 
 from rolo import Request
 from rolo.gateway import RequestContext
+from werkzeug.datastructures import Headers
 
+from localstack.aws.api.apigateway import Method, Resource
 from localstack.services.apigateway.models import RestApiDeployment
 
 
+class InvocationRequest(TypedDict, total=False):
+    http_method: Optional[HTTPMethod]
+    """HTTP Method of the incoming request"""
+    raw_path: Optional[str]
+    """Raw path of the incoming request with no modification, needed to keep double forward slashes"""
+    path: Optional[str]
+    """Path of the request with no URL decoding"""
+    path_parameters: Optional[dict[str, str]]
+    """Path parameters of the request"""
+    query_string_parameters: Optional[dict[str, str]]
+    """Query string parameters of the request"""
+    # TODO: need to check if we need the raw headers (as it's practical for casing reasons)
+    raw_headers: Optional[Headers]
+    """Raw headers using the Headers datastructure which allows access with no regards to casing"""
+    headers: Optional[dict[str, str]]
+    """Headers of the request"""
+    multi_value_query_string_parameters: Optional[dict[str, list[str]]]
+    """Multi value query string parameters of the request"""
+    multi_value_headers: Optional[dict[str, list[str]]]
+    """Multi value headers of the request"""
+    body: Optional[bytes]
```

Comment:
Would it be useful to have `raw_body` and `decoded_body`, or similar?

Reply:
Not sure I understand, how do you mean?

Currently, we use `restore_payload` for the body, because the request parsing of Werkzeug might consume the data to create a parsed form, and then we would not be able to forward/read the raw body. 

It splits in 3 parts: `data`, `form` and `files`. My belief is that APIGW does not manipulate different parts of the body directly? But I'm not sure. 

It's also possible I'm not familiar with what is possible to do with APIGW 😅

Reply:
:thinking: I think the only thing we can do on the body is through the request templates. I haven't looked fully into the integration, so maybe this is the right way to go. I was just concerned about decoding it in with LS parser to restore it, to decode it again to apply the templates.

Reply:
Ah, I see. 
I think the issue here is what `restore_payload` does. We don't really "decode", we "re-encode" it to be like what we actually received from the client. 

For example, the usual way is to call `request.data` to get the `bytes` content of the body. 

But let's imagine your request is a `Content-Type: application/x-www-form-urlencoded`.
You get an HTTP body looking like this: `field1=value1&field2=value2`

When accessing `request.data`, you'll get... `b""`, an empty `bytes` object because Werkzeug will have parsed the request and consumed the stream, and set `request.form` to be equal to an `ImmutableMultiDict` looking like `{"field1": "value1", "field2": "value2"}`.

One of the way here is to directly consume the `request.stream` once to get the raw data out. But anything earlier in the handler chain called `request.data`, the stream will have be consumed and you have no way to get it back (yet! with the new `twisted` web server, we actually have a way now 🥳 which means we can have a short circuit `restore_payload`), which is why we re-encode the body 😕 

So in itself I'd say we have the proper `raw_body` now, the one the request template should act upon?

Reply:
I guess it makes sense then since only one handler should need to perform operations on the body, that step can ensure the body returns to bytes once it has completed and and shouldn't need to keep a dict representation of it anywhere. :+1: